File: Depth-First Search/Binary tree/Find paths/Medium/LET_no1120_Maximum Average Subtree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maximumAverageSubtree(self, root: Optional[TreeNode]) -> float:
        res, _, _ = self.helper(root)
        return res
    
    def helper(self, root):
        if not root:
            return 0, 0, 0
        
        leftRes, leftSum, leftCount = self.helper(root.left)
        rightRes, rightSum, rightCount = self.helper(root.right)
        
        curSum = leftSum + rightSum + root.val
        curCount = leftCount + rightCount + 1
        curRes = curSum / curCount
        
        return max(leftRes, rightRes, curRes), curSum, curCount

File: Depth-First Search/Binary tree/Find paths/Medium/LET_no1120_Maximum Average Subtree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    double res = 0.0;
public:
    double maximumAverageSubtree(TreeNode* root) {
        dfs(root);
        return res;
    }

    pair<int,int> dfs(TreeNode* root) {
        if (!root)
            return {0, 0};

        pair<int, int> left_res = dfs(root->left);
        pair<int, int> right_res = dfs(root->right);
        int sum = left_res.first + right_res.first+ root->val;
        int count = left_res.second + right_res.second + 1;
        
        res = max(res, double(sum) / count);
        
        return {sum, count};
    }
};